presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PruneUnreferencedOutputs.java

```diff
@@ -812,11 +817,25 @@ public PlanNode visitAssignUniqueId(AssignUniqueId node, RewriteContext<Set<Symb
         @Override
         public PlanNode visitLateralJoin(LateralJoinNode node, RewriteContext<Set<Symbol>> context)
         {
-            PlanNode subquery = context.rewrite(node.getSubquery(), context.get());
+            Set<Symbol> expectedFilterSymbols = SymbolsExtractor.extractUnique(node.getFilter());
+
+            Set<Symbol> expectedFilterAndContextSymbols = ImmutableSet.<Symbol>builder()
+                    .addAll(expectedFilterSymbols)
+                    .addAll(context.get())
+                    .build();
+
+            PlanNode subquery = context.rewrite(node.getSubquery(), expectedFilterAndContextSymbols);
 
             // remove unused lateral nodes
-            if (intersection(ImmutableSet.copyOf(subquery.getOutputSymbols()), context.get()).isEmpty() && isScalar(subquery)) {
-                return context.rewrite(node.getInput(), context.get());
+            if (intersection(ImmutableSet.copyOf(subquery.getOutputSymbols()), context.get()).isEmpty()) {
+                // remove unused lateral subquery of inner join
+                if (node.getType() == INNER && isScalar(subquery) && node.getFilter().equals(TRUE_LITERAL)) {
```

Comment:
we could also check if subquery cardinality is 0, then we could optimize it to empty `ValuesNode`.

Also I think we should extract this to optimizer Rule and remove this code from here.